ports/arrow/portfile.cmake

```
@@ -3,53 +3,72 @@ vcpkg_fail_port_install(ON_ARCH "x86" "arm" "arm64")
     HEAD_REF master
     PATCHES
         all.patch
 )
 
-string(COMPARE EQUAL ${VCPKG_LIBRARY_LINKAGE} "dynamic" ARROW_BUILD_SHARED)
-string(COMPARE EQUAL ${VCPKG_LIBRARY_LINKAGE} "static" ARROW_BUILD_STATIC)
-
 vcpkg_check_features(OUT_FEATURE_OPTIONS FEATURE_OPTIONS
-    "csv"         ARROW_CSV
-    "json"        ARROW_JSON
-    "parquet"     ARROW_PARQUET
-    "filesystem"  ARROW_FILESYSTEM
+    FEATURES
+        csv         ARROW_CSV
+        dataset     ARROW_DATASET
+        filesystem  ARROW_FILESYSTEM
+        flight      ARROW_FLIGHT
+        json        ARROW_JSON
+        orc         ARROW_ORC
+        parquet     ARROW_PARQUET
+        parquet     PARQUET_REQUIRE_ENCRYPTION
+        s3          ARROW_S3
 )
 
-file(REMOVE "${SOURCE_PATH}/cpp/cmake_modules/FindZSTD.cmake")
+if(VCPKG_TARGET_IS_WINDOWS OR VCPKG_TARGET_IS_UWP)
+    set(MALLOC_OPTIONS -DARROW_JEMALLOC=OFF)
+elseif("jemalloc" IN_LIST FEATURES)
+    set(MALLOC_OPTIONS -DARROW_JEMALLOC=ON)
+else()
+    set(MALLOC_OPTIONS -DARROW_JEMALLOC=OFF)
+endif()
+
+if(VCPKG_TARGET_IS_WINDOWS AND ("mimalloc" IN_LIST FEATURES))
+    set(MALLOC_OPTIONS ${MALLOC_OPTIONS} -DARROW_MIMALLOC=ON)
+else()
+    set(MALLOC_OPTIONS ${MALLOC_OPTIONS} -DARROW_MIMALLOC=OFF)
+endif()
+
+string(COMPARE EQUAL ${VCPKG_LIBRARY_LINKAGE} "dynamic" ARROW_BUILD_SHARED)
+string(COMPARE EQUAL ${VCPKG_LIBRARY_LINKAGE} "static" ARROW_BUILD_STATIC)
+string(COMPARE EQUAL ${VCPKG_LIBRARY_LINKAGE} "dynamic" ARROW_DEPENDENCY_USE_SHARED)
+
+if(VCPKG_TARGET_IS_WINDOWS)
+    set(THRIFT_USE_SHARED OFF)
+else()
+    set(THRIFT_USE_SHARED ${ARROW_DEPENDENCY_USE_SHARED})
+endif()
 
 vcpkg_configure_cmake(
     SOURCE_PATH ${SOURCE_PATH}/cpp
     PREFER_NINJA
     OPTIONS
-        -DARROW_DEPENDENCY_SOURCE=SYSTEM
-        -Duriparser_SOURCE=SYSTEM
-        -DARROW_BUILD_TESTS=OFF
         ${FEATURE_OPTIONS}
-        -DARROW_BUILD_STATIC=${ARROW_BUILD_STATIC}
+        ${MALLOC_OPTIONS}
         -DARROW_BUILD_SHARED=${ARROW_BUILD_SHARED}
-        -DARROW_BROTLI_USE_SHARED=${ARROW_BUILD_SHARED}     # This can be wrong in custom triplets
-        -DARROW_GFLAGS_USE_SHARED=${ARROW_BUILD_SHARED}     # This can be wrong in custom triplets
-        -DARROW_LZ4_USE_SHARED=${ARROW_BUILD_SHARED}        # This can be wrong in custom triplets
-        -DARROW_SNAPPY_USE_SHARED=${ARROW_BUILD_SHARED}     # This can be wrong in custom triplets
-        -DARROW_THRIFT_USE_SHARED=OFF                       # vcpkg doesn't build Thrift as a shared library for the moment (2020/01/22).
-        -DARROW_UTF8PROC_USE_SHARED=${ARROW_BUILD_SHARED}   # This can be wrong in custom triplets
-        -DARROW_ZSTD_USE_SHARED=${ARROW_BUILD_SHARED}       # This can be wrong in custom triplets
-        -DARROW_JEMALLOC=OFF
-        -DARROW_BUILD_UTILITIES=OFF
+        -DARROW_BUILD_STATIC=${ARROW_BUILD_STATIC}
+        -DARROW_BUILD_TESTS=OFF
+        -DARROW_DEPENDENCY_SOURCE=SYSTEM
+        -DARROW_DEPENDENCY_USE_SHARED=${ARROW_DEPENDENCY_USE_SHARED}
```

```suggestion
        -DARROW_DEPENDENCY_USE_SHARED=OFF
```

Arrow cannot assume that its dependencies are built with any particular linkage -- instead, we should pass a constant here (ON or OFF) and patch the buildsystem to be agnostic.